### PR TITLE
expression: implement vectorized evaluation for `builtinPeriodDiffSig`

### PR DESCRIPTION
**Pull request by shihongzhi**
### What problem does this PR solve? 

implement vectorized evaluation for builtinPeriodDiffSig , for #12103

### What is changed and how it works?

	$go test -v -benchmem -bench=BenchmarkVectorizedBuiltinTimeFunc -run=BenchmarkVectorizedBuiltinTimeFunc -args "builtinPeriodDiffSig"
	goos: darwin
	goarch: amd64
	pkg: github.com/pingcap/tidb/expression
	BenchmarkVectorizedBuiltinTimeFuncGenerated-4   	1000000000	         0.0116 ns/op	       0 B/op	       0 allocs/op
	BenchmarkVectorizedBuiltinTimeFunc/builtinPeriodDiffSig-VecBuiltinFunc-4         	  107150	     11805 ns/op	       0 B/op	       0 allocs/op
	BenchmarkVectorizedBuiltinTimeFunc/builtinPeriodDiffSig-NonVecBuiltinFunc-4      	   36326	     31862 ns/op	       0 B/op	       0 allocs/op
	PASS
	ok  	github.com/pingcap/tidb/expression	2.970s

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: expression/builtin_time_vec.go [hunk 1]
```diff
@@ -313,11 +313,49 @@ func (b *builtinSubDateStringDecimalSig) vecEvalTime(input *chunk.Chunk, result
...
+		return err
+	}
+	defer b.bufAllocator.put(buf1)
+	if err := b.args[1].VecEvalInt(b.ctx, input, buf1); err != nil {
+		return err
+	}
+
+	period1s := buf.Int64s()
+	period2s := buf1.Int64s()
+	result.ResizeInt64(n, false)
```
Comment (SunRunAway) [inline expression/builtin_time_vec.go]: We can reuse `result` buffer with `buf0` or `buf1`
Here's an [example](https://github.com/pingcap/tidb/blob/8de9c26805c5eb4be80fb3268180418a6c20e3b1/expression/builtin_math_vec.go#L631).

File: expression/builtin_time_vec_test.go [hunk 2]
```diff
@@ -52,11 +52,13 @@ var vecBuiltinTimeCases = map[string][]vecExprBenchCase{
...
-	ast.MakeDate:   {},
-	ast.MakeTime:   {},
-	ast.PeriodAdd:  {},
-	ast.PeriodDiff: {},
-	ast.Quarter:    {},
+	ast.MakeDate:  {},
+	ast.MakeTime:  {},
+	ast.PeriodAdd: {},
+	ast.PeriodDiff: {
+		{retEvalType: types.ETInt, childrenTypes: []types.EvalType{types.ETInt, types.ETInt}, geners: []dataGenerator{&rangeInt64Gener{1, 13}, &rangeInt64Gener{1, 13}}},
```
Comment (SunRunAway) [inline expression/builtin_time_vec_test.go]: Could you make a new `gener` in builtin_time_vec_test.go?

File: expression/builtin_time_vec.go [hunk 3]
```diff
@@ -313,11 +313,49 @@ func (b *builtinSubDateStringDecimalSig) vecEvalTime(input *chunk.Chunk, result
...
+		if buf.IsNull(i) || buf1.IsNull(i) {
+			result.SetNull(i, true)
+			continue
+		}
+		if !validPeriod(period1s[i]) || !validPeriod(period2s[i]) {
+			if err := errIncorrectArgs.GenWithStackByArgs("period_diff"); err != nil {
+				return err
+			}
+			i64s[i] = 0
+			continue
```
Comment (SunRunAway) [inline expression/builtin_time_vec.go]: The `err` will never be `nil` here.

File: expression/builtin_time_vec_test.go (hunk 2 above)
Comment (shihongzhi) [reply expression/builtin_time_vec_test.go]: Yes, I add `periodGener`

**Pull request merged**